sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateTable.scala

```diff
@@ -26,19 +26,25 @@ import org.apache.spark.sql.connector.read.ScanBuilder
 import org.apache.spark.sql.execution.datasources.v2.state.StateSourceOptions.JoinSideValues
 import org.apache.spark.sql.execution.datasources.v2.state.utils.SchemaUtil
 import org.apache.spark.sql.execution.streaming.TransformWithStateVariableInfo
-import org.apache.spark.sql.execution.streaming.state.{KeyStateEncoderSpec, StateStoreColFamilySchema, StateStoreConf}
+import org.apache.spark.sql.execution.streaming.state.{KeyStateEncoderSpec, StateSchemaProvider, StateStoreColFamilySchema, StateStoreConf}
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
 
-/** An implementation of [[Table]] with [[SupportsRead]] for State Store data source. */
+/**
+ * An implementation of [[Table]] with [[SupportsRead]] for State Store data source.
+ * @param stateSchemaProviderOpt Optional provider that maintains mapping between schema IDs and
```

Comment:
One question: what would be the schema of the state when we do time-travel? And what would be the schema of the state when we read from changelog in version range?

Reply:
So the schema(s) we would use for a requested committed batch are the most recent schema(s) (in reverse chronological order) that are noted within the operator metadata. So, in this case, for the following scenarios:
- using latest snapshot
- using specified snapshot
- using changelog from startVersion to endVersion

we would have all the required schema information available since the evolutions are only incremental as of now (when we do support full state rewrite with latest schema - this will not hold true and we can only use the reader between/within full rewrites - not across rewrites)